engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/ConfluentAvroToJsonFormatterFactory.scala

```diff
@@ -0,0 +1,121 @@
+      messageFormatter.getSchemaIdOpt(record.value()).get,
+      SerializableConsumerRecord(deserializedRecord)
+    )
+
+    consumerRecordEncoder(serializableRecord).noSpaces.getBytes(StandardCharsets.UTF_8)
+  }
+
+  /**
+    * Step 1: Deserialize raw json bytes to AvroSerializableConsumerRecord[Json, Json] domain without interpreting key and value content.
+    * Step 2: Create key and value json-to-avro interpreter based on schema id's provided in json.
+    * Step 3: Use interpreter to create raw kafka ConsumerRecord
+    */
+  override protected def parseRecord(topic: String, bytes: Array[Byte]): ConsumerRecord[Array[Byte], Array[Byte]] = {
+    val record = CirceUtil.decodeJsonUnsafe(bytes)(consumerRecordDecoder)
+
+    def serializeKeyValue(keyOpt: Option[Json], value: Json): (Array[Byte], Array[Byte]) = {
+      val reader = createReader(topic)
+      val keyBytes = if (kafkaConfig.useStringForKey) {
+        keyOpt match {
+          // we handle strings this way because we want to keep result value compact and JString is formatted in quotes
+          case Some(j) if j.isString => j.asString.get.getBytes(StandardCharsets.UTF_8)
+          case Some(j) => j.noSpaces.getBytes(StandardCharsets.UTF_8)
+          case None => null
+        }
+      } else {
+        val keySchema = record.keySchemaId.map(id => reader.schemaById(id)).getOrElse(throw new IllegalArgumentException("Error reading schema: empty schema id"))
+        keyOpt.map(keyJson => reader.readJson[K](keyJson, Option(keySchema), ConfluentUtils.keySubject(topic))).getOrElse(Array.emptyByteArray)
+      }
+      val valueSchema = reader.schemaById(record.valueSchemaId)
+      val valueBytes = reader.readJson[V](value, Option(valueSchema), ConfluentUtils.valueSubject(topic))
+      (keyBytes, valueBytes)
+    }
+    record.consumerRecord.toKafkaConsumerRecord(topic, serializeKeyValue)
+  }
+
+  override protected def testDataSplit: TestDataSplit = TestParsingUtils.newLineSplit
+
+  implicit protected val serializableRecordDecoder: Decoder[SerializableConsumerRecord[Json, Json]] = deriveDecoder
+  protected val consumerRecordDecoder: Decoder[AvroSerializableConsumerRecord[Json, Json]] = deriveDecoder
+
+  implicit protected val keyEncoder: Encoder[K] = createKeyEncoder[K](kafkaConfig.useStringForKey, messageFormatter)
+  implicit protected val valueEncoder: Encoder[V] = createValueEncoder[V](messageFormatter)
+  implicit protected val serializableRecordEncoder: Encoder[SerializableConsumerRecord[K, V]] = deriveEncoder
+  protected val consumerRecordEncoder: Encoder[AvroSerializableConsumerRecord[K, V]] = deriveEncoder
+
+}
+
+object ConfluentAvroToJsonFormatter {
+
+  def createKeyEncoder[K: ClassTag](useStringForKey: Boolean, messageFormatter: ConfluentAvroMessageFormatter): Encoder[K] = {
+    new Encoder[K] {
+      override def apply(key: K): Json = if (useStringForKey) {
+        Json.fromString(key.asInstanceOf[String])
```

Comment:
Is this code is correct? Shouldn't `serializeKeyValue(encoder.apply(key), ...)` always return `key`? If so, correct would be just `key.asInstanceOf[String]` ... I wonder if we can have `serializeKey` near `encode` and the same for `value`. Maybe it could be in the class instead of companion object?

Reply:
Fixed